### ISSUE TITLE
Custom triton kernels with autotune running in torch.compile() don't seem to properly capture mutating inputs

### ISSUE THREAD

**Issue by ani300** [pytorch/pytorch#120802]
### 🐛 Describe the bug

While debugging some accuracy issues with a custom triton kernel with autotune configs running inside a torch.compile() environment, I've noticed that when the kernel has a few configs available defined through `@triton.autotune()`, the results will be wrong for the first run (when autotune is supposed to run).

While digging deep into how these kernels get put into the inductor codegen path, I've noticed that while inductor-generated kernels have `mutated_args` as part of their `inductor_meta` properties, custom kernels lack this parameter, which `cached_autotune` will use to prevent the issue I'm talking about from happening.

If I understand the rest of the code correctly, this seems to stem from the custom triton kernel getting functionalized in the AOT autograd step, so maybe the issue is there. In any case, I'll try to create a small repro if this is not enough info, as the current code where this happens is not easy to ship.

### Versions

Collecting environment information...
PyTorch version: 2.3.0.dev20240226
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-1041-aws-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB
GPU 4: NVIDIA A100-SXM4-80GB
GPU 5: NVIDIA A100-SXM4-80GB
GPU 6: NVIDIA A100-SXM4-80GB
GPU 7: NVIDIA A100-SXM4-80GB

Nvidia driver version: 535.54.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          96
On-line CPU(s) list:             0-95
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) Platinum 8275CL CPU @ 3.00GHz
Stepping:                        7
CPU MHz:                         2999.998
BogoMIPS:                        5999.99
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       1.5 MiB
L1i cache:                       1.5 MiB
L2 cache:                        48 MiB
L3 cache:                        71.5 MiB
NUMA node0 CPU(s):               0-23,48-71
NUMA node1 CPU(s):               24-47,72-95
Vulnerability Itlb multihit:     KVM: Mitigation: VMX unsupported
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Mmio stale data:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke

Versions of relevant libraries:
[pip3] mypy==1.7.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.2
[pip3] torch==2.3.0.dev20240226
[pip3] torchaudio==2.2.0.dev20240226
[pip3] torchvision==0.18.0.dev20240226
[pip3] triton==3.0.0
[conda] blas                      2.116                       mkl    conda-forge
[conda] blas-devel                3.9.0            16_linux64_mkl    conda-forge
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            16_linux64_mkl    conda-forge
[conda] mkl                       2022.1.0           h84fe81f_915    conda-forge
[conda] mkl-devel                 2022.1.0           ha770c72_916    conda-forge
[conda] mkl-include               2022.1.0           h84fe81f_915    conda-forge
[conda] numexpr                   2.8.3           mkl_py311h0b1cd97_1    conda-forge
[conda] numpy                     1.26.2          py311h64a7726_0    conda-forge
[conda] pytorch                   2.3.0.dev20240226 py3.11_cuda12.1_cudnn8.9.2_0    pytorch-nightly
[conda] pytorch-cuda              12.1                 ha16c6d3_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.2.0.dev20240226     py311_cu121    pytorch-nightly
[conda] torchtriton               3.0.0+901819d2b6           py311    pytorch-nightly
[conda] torchvision               0.18.0.dev20240226     py311_cu121    pytorch-nightly

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @oulgen

**Comment by ani300** [comment]
A repro for this particular issue is the following:

```
import torch
import triton
import triton.language as tl


@triton.autotune(configs=[
    triton.Config(kwargs={'BLOCK_SIZE': 16}, num_warps=8, num_stages=4),
    triton.Config(kwargs={'BLOCK_SIZE': 32}, num_warps=8, num_stages=4),
  ],
  key=['n_elements'],
  reset_to_zero=["output_ptr"],
)
@triton.jit
def add_kernel(x_ptr,  # *Pointer* to first input vector.
               y_ptr,  # *Pointer* to second input vector.
               output_ptr,  # *Pointer* to output vector.
               n_elements,  # Size of the vector.
               BLOCK_SIZE: tl.constexpr,  # Number of elements each program should process.
               # NOTE: `constexpr` so it can be used as a shape value.
               ):
    # There are multiple 'programs' processing different data. We identify which program
    # we are here:
    pid = tl.program_id(axis=0)  # We use a 1D launch grid so axis is 0.
    # This program will process inputs that are offset from the initial data.
    # For instance, if you had a vector of length 256 and block_size of 64, the programs
    # would each access the elements [0:64, 64:128, 128:192, 192:256].
    # Note that offsets is a list of pointers:
    block_start = pid * BLOCK_SIZE
    offsets = block_start + tl.arange(0, BLOCK_SIZE)
    # Create a mask to guard memory operations against out-of-bounds accesses.
    mask = offsets < n_elements
    # Load x and y from DRAM, masking out any extra elements in case the input is not a
    # multiple of the block size.
    x = tl.load(x_ptr + offsets, mask=mask)
    y = tl.load(y_ptr + offsets, mask=mask)
    output = x + y
    # Write x + y back to DRAM.
    tl.atomic_add(output_ptr + offsets, output, mask=mask)


def add(x: torch.Tensor, y: torch.Tensor):
    # We need to preallocate the output.
    output = torch.zeros_like(x)
    assert x.is_cuda and y.is_cuda and output.is_cuda
    n_elements = output.numel()
    # The SPMD launch grid denotes the number of kernel instances that run in parallel.
    # It is analogous to CUDA launch grids. It can be either Tuple[int], or Callable(metaparameters) -> Tuple[int].
    # In this case, we use a 1D grid where the size is the number of blocks:
    grid = lambda meta: (triton.cdiv(n_elements, meta['BLOCK_SIZE']), )
    # NOTE:
    #  - Each torch.tensor object is implicitly converted into a pointer to its first element.
    #  - `triton.jit`'ed functions can be indexed with a launch grid to obtain a callable GPU kernel.
    #  - Don't forget to pass meta-parameters as keywords arguments.
    add_kernel[grid](x, y, output, n_elements)
    # We return a handle to z but, since `torch.cuda.synchronize()` hasn't been called, the kernel is still
    # running asynchronously at this point.
    return output


def repro_func(x, y):
    return add(x, y)


x = torch.randn((200,), device="cuda")
y = torch.randn((200,), device="cuda")

compiled_repro = torch.compile(repro_func)

compile1_sum = compiled_repro(x, y)
compile2_sum = compiled_repro(x, y)


torch.testing.assert_close(compile2_sum, compile1_sum)
```

This outputs the following:

```
/lustre/aviros/conda_envs/fms-dev/lib/python3.11/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
Traceback (most recent call last):
  File "/lustre/aviros/fms-oss/repro2.py", line 73, in <module>
    torch.testing.assert_close(compile2_sum, compile1_sum)
  File "/lustre/aviros/conda_envs/fms-dev/lib/python3.11/site-packages/torch/testing/_comparison.py", line 1523, in assert_close
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 200 / 200 (100.0%)
Greatest absolute difference: 4064.283203125 at index (95,) (up to 1e-05 allowed)
Greatest relative difference: 0.9989496469497681 at index (6,) (up to 1.3e-06 allowed)
```

**Comment by oulgen** [comment]
@ani300 we do not plan to support "reset_to_zero=["output_ptr"]" feature. I will write a PR to hard error on this.

**Comment by ani300** [comment]
ok! good to know about the particular setting. I was originally using this to pass num_warps and num_stages to the kernel as passing them as parameters is not working either unless you use the autotuner configs. Is this going to be supported?

**Comment by oulgen** [comment]
Can you show what you mean by passing them as parameters? It should work when you use them as autotuner configs as you mentioned.

**Comment by ani300** [comment]
something like:
```
add_kernel[grid](x, y, output, n_elements, num_warps=8, num_stages=4)
```
keeping the rest of the code intact (and with no autotuner). This is supported when running in eager mode (or running the kernel inside a torch.library defined op which makes it skip the compile path) but breaks compile

**Comment by oulgen** [comment]
Just looked at what's happening, and it appears the reason why it is not supported is because inductor's custom precompile/autotune step. It should not be too hard to add it to somewhere here: https://github.com/pytorch/pytorch/blob/797d4fbdf423dd9320ebe383fb57ffb1135c4a99/torch/_inductor/triton_heuristics.py#L404-L419

If you're up for it, feel free to send a PR updating inductor. Otherwise, since you can get around this via `triton.autotune`, this will likely become backlog for me.

**Comment by ani300** [comment]
I'll add it to my backlog as well, as I don't have the bandwidth for it right now, but it's going to be eventually nice to have. For now, I have the kernel running inside a torch.ops wrapper, which is enough for what I need.
